### PR TITLE
HDFS-15594. Lazy calculate live datanodes in safe mode tip

### PR DESCRIPTION
**Pull request by NickyYe**
https://issues.apache.org/jira/browse/HDFS-15594

## NOTICE

Please create an issue in ASF JIRA before opening a pull request,
and you need to set the title of the pull request which starts with
the corresponding JIRA issue number. (e.g. HADOOP-XXXXX. Fix a typo in YYY.)
For more details, please see https://cwiki.apache.org/confluence/display/HADOOP/How+To+Contribute

### CODE REVIEW

**Review by Hexiaoqiao — APPROVED** [review]
Thanks @NickyYe for your works, +1 from my side.
cc @goiri would you mind to have another check?

**Review by goiri — APPROVED** [review]

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManagerSafeMode.java [hunk 1]
```diff
@@ -309,16 +311,21 @@ String getSafeModeTip() {
...
+        } else {
+          msg += String.format("The number of live datanodes %d has reached "
+                  + "the minimum number %d. ",
+              numLive, datanodeThreshold);
+        }
       } else {
-        msg += String.format("The number of live datanodes %d has reached "
-                + "the minimum number %d. ",
-            numLive, datanodeThreshold);
+        msg += "The number of live datanodes is not calculated " +
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManagerSafeMode.java:327]: As we are at it, does it make sense to use StringBuilder?

Comment (NickyYe) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManagerSafeMode.java]: fixed.

**Comment by goiri** [comment]
Thanks @NickyYe for the StringBuilder, that will reduce some memory too.
I'm a little surprised that Yetus is not kicking in.
@ayushtkn are you familiar on how to kick it? And now that I have your attention... any issues you can see with changing the message.

**Review by ayushtkn — APPROVED** [review]
I have triggered the build -
https://ci-hadoop.apache.org/job/hadoop-multibranch/view/change-requests/job/PR-2332/4/

The improvement seems fair enough, I don't think the message change would create any problem, Changing messages at UI is allowed.
It would change the exception text, I guess when there is a `SafeModeException` through `checkNameNodeSafeMode` in one case, but I don't think that would bother anyone as such. Should be all good.

**Comment by goiri** [comment]
There were a lot of failures because TestFileChecksum:
https://ci-hadoop.apache.org/job/hadoop-multibranch/view/change-requests/job/PR-2332/4/testReport/
But they are not related.

**Pull request merged**